Relax atomic orderings + fix parking/unparking UB

Cassy343 opened the pull request:
I found this crate when looking for a simple oneshot implementation and decided to try to make some improvements. I removed the UB caused by taking multiple mutable references to the channel concurrently, and fixed a couple UAF errors which could crop up due to unlucky thread interleaving (specifically if the sending thread yielded while unparking the receiving thread).

I'm making this a draft pull request since I'm not very satisfied with the implementation of `Future::poll`. I think there are some improvements that could be made there and I have not spent as much mental energy verifying its correctness. I noticed that the receiver itself currently implements `Future`, but I feel like it would be a lot more ergonomic to add a method called `into_future` which would then return something you could await. This would be a breaking API change, but would give you a compile-time guarantee that the receiver could never be used in a sync context once polled asynchronously.

I'd love to hear your thoughts.

Should close #10

faern commented:
Oh, there is UB here? I have too look closer at that, but this is a pretty large PR. Could you possibly elaborate further on how the UB can be triggered?

I like the idea of moving the `impl Future` to a separate type to guarantee it's not being synchronously received after async polling started!

Cassy343 commented:
The first form of UB here is the multiple mutable reference issue. This is triggered if one thread calls `recv` or some other method on the receiver at the same time that another thread calls `send` on the sender. When this happens each thread has a mutable reference to the same memory location which is defined to be instant UB. This PR fixes that by only using shared references and going through `UnsafeCell`.

The second issue was extremely subtle, and occurs under the following scenario:
- Receiving thread begins receiving on an empty channel
- Sending thread writes the message to the channel, sets the state to `MESSAGE`, and gets pre-empted before waking the receiving thread
- Receiving thread is spuriously unparked, sees the state is set to `MESSAGE`, takes the message and deallocates the channel in memory.
- Sending thread resumes execution, and attempts to read the waker from the freed channel
This issue was resolved by adding the intermediate `UNPARKING` state. I was able to reproduce this in valgrind by inserting a liberal amount of `std::thread::yield_now()`s in inconvenient places.

faern commented:
Thanks for the explanation. I'll try to take the time to dive deeper into this PR later in the week, and next week!

> I noticed that the receiver itself currently implements Future, but I feel like it would be a lot more ergonomic to add a method called into_future which would then return something you could await

Yes, this sounds like a very good idea! And the soon-to-be-stable `IntoFuture` trait seems to solve this beautifully. Then the `Receiver` can be `await`ed directly, but it's still a separate type. Until then, and maybe as a long term alternative to relying on this trait, we can provide an explicit method. IMO `Receiver<T>::recv_async(self) -> AsyncReceiver<T>` sounds like good names. This is a great API improvement as it can remove possible panics from the existing sync `recv*` calls :partying_face: 

This PR touches on many more or less unrelated subjects at once. Such as deduplicating the crate documentation and a few other things. If possible, I would prefer to have this split into multiple PRs. Mostly to simplify reviewing, and get stuff merged in bite size chunks.

Cassy343 commented:
Sounds good. Once you get a feel for the changes made (to be honest most of them are just de-duplicating code, changing code to use more up-to-date unsafe patterns that are the same at run-time, and comments), let me know what chunks you'd like to see this split into and I can get on that when I have time. Once that is done I'll tackle the async API change.

faern commented:
I have now reviewed the entire first commit in this PR (https://github.com/faern/oneshot/pull/17/commits/e1d5fbde4609a2c16584dfcf10c95a99c45ae633). I think it looks correct, except the miss with `Ok` vs `Err` on the conversion to `compare_exchange` in a bunch of places.

Since the `compare_and_swap` -> `compare_exchange` conversion is not needed to fix the UB that the commit is otherwise dealing with, may I suggest that you move that conversion to a separate commit?

Thank you a lot for this PR! Even if I have some feedback on the PR format, I deeply appreciate that you have found and fixed these UBs!

Cassy343 commented:
I'm working on rewriting the history a bit right now to make it more clear. The first commit is basically the same but with that `Ok`/`Err` issue fixed. I changed the chunky commit that fixes the UB to *just* contain code changes, and I removed the documentation changes since that can be handled after the code changes are settled.

Cassy343 commented:
Okay I think I did that correctly. I'm not super great with git so if that really screwed things up let me know and I'll do my best to fix it!

> Thank you a lot for this PR! Even if I have some feedback on the PR format, I deeply appreciate that you have found and fixed these UBs!

It's my pleasure! :)

faern commented:
Thanks. However, I see that you put the change to `UnsafeCell` and `NonNull`, which I had reviewed, together with all the new atomic operations, which I have not reviewed. The former are way simpler code changes and are fixing the first and most glaring UB you found. Can we focus on having just that in one commit, so I can merge that to master before we start looking at the intricate atomics changes?

A reason for having them separate, apart from being easier to review, is that it becomes easier to bisect bugs. If a single commit changes all of this at once and we later find a bug, it's going to be harder to pinpoint which code change caused it, if it's all in one commit.

faern commented on src/errors.rs:
```diff
@@ -1,50 +1,62 @@
-        let channel: &mut super::Channel<T> = unsafe { &mut *self.channel_ptr };
+        let channel_ptr = self.channel_ptr;
+        // SAFETY: we have ownership of the channel
+        let channel: &Channel<T> = unsafe { channel_ptr.as_ref() };
```
This is a common pattern now in the code with your change. To be a bit nitpicky I think `let channel` should be created after `forget(self)`. Since this statement is no longer depending on `self` it has no technical reason to be before it. And it's good practice to declare variables as close to where they are used as possible.

As I said, it's a bit nitpicky, but I think the following order would be cleaner:

```rust
let channel_ptr = self.channel_ptr;
mem::forget(self);

let channel: &Channel<T> = unsafe { channel_ptr.as_ref() };
// Do stuff with `channel`
```

Cassy343 replied:
Resolved!

faern commented on src/errors.rs:
```diff
@@ -1,50 +1,62 @@
         mem::forget(self);
 
+        // SAFETY: we know that the message is initialized according to the safety requirements of
+        // `new`
```
The documentation for `new` does not say that the message field must be initialized. It only talks about it transferring ownership.

faern commented:
Thanks for the improved commits! I had forgotten to approve the CI to run on this PR. Now when I did that I see that there are both `rustfmt` formatting to fix. And also, you use code that was stabilized after the current MSRV (1.56). It's fine to bump the MSRV, but it should be noted in the changelog, and the github actions workflow file needs to be updated also.

I added an extra CI job to also run Clippy. The CI on this repository was a bit lacking, and a bit outdated. Could you please cater to the current complaints of the CI and rebase on latest `master` where I have added Clippy?

faern commented:
I cherry-picked your commit on the migration to `compare_exchange`. The commit was pretty trivial, it was your first commit, and it made Clippy happy. So that part of this PR is now "merged" already :) You can drop that commit when you rebase on latest `master`.

Cassy343 commented:
I fixed the formatting and clippy issue. Also it's definitely not needed to bump the MSRV for `MaybeUninit::assume_init_drop` since that can be pretty trivially re-implemented, so that's what I did. I also rebased on the latest master and dropped that first commit.

As always don't hesitate to point things out as you go through and review these changes!

faern commented:
Can you maybe squash [Implement latest feedback](https://github.com/faern/oneshot/pull/17/commits/8a1cfd7014804e99658a59ce34b1f1397b05881b) and [Deduplicate dealloc code](https://github.com/faern/oneshot/pull/17/commits/700675ce10d781ed3553a7c267dcd23ac68227dc) into [Improve unsafe hygiene](https://github.com/faern/oneshot/pull/17/commits/409c49879db717bd9569326bb7d0123c2bf54c1f) ? As pointed out earlier, I would like to be able to get the UB fix regarding multiple mutable references merged in one go, without all the changes to the atomic orderings and other fixes. That commit will also need a `rustfmt` run. And it also has Clippy errors.

faern commented on src/errors.rs:
```diff
@@ -1,50 +1,64 @@
+    /// channel's resources to the created `SendError`. Thus the caller must ensure that the
+    /// pointer is not used in a way which would violate this ownership transfer. Moreover, the
+    /// channel must contain an initialized message.
+    pub(crate) const unsafe fn new(channel_ptr: NonNull<Channel<T>>) -> Self {
```
I think we can consider making this `const unsafe fn new(channel: Box<Channel<T>>`. It would still be unsafe, since the caller must make sure the `message` is initialized, but that would remove a lot of the unsafety in this file regarding ownership and who frees the memory. The box going out of scope would automatically free the memory.

faern replied:
I'm having second thoughts on this. In some future I want to investigate making this crate not have to depend on `alloc`, and allow it to work by having the user supply the storage for the `Channel`. With that in mind I don't want the error to have a stronger dependency on `Box` than it already has. Keep this as is!

Cassy343 replied:
Will do!

faern commented on src/lib.rs:
```diff
@@ -186,23 +198,54 @@ pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
 #[derive(Debug)]
 pub struct Sender<T> {
-    channel_ptr: *mut Channel<T>,
+    channel: NonNull<Channel<T>>,
```
It is inconsistent to name the field `channel_ptr` in `SendError` but name it `channel` here when they currently are the exact same type. The inconsistency is spreading with a lot of code currently looking like `let channel_ptr = self.channel;`. I would love if that variable was named the same as the field. However, see my other comment on potentially changing the type in `SendError` :)

I think we should name the field `channel_ptr` if it contains a `NonNull<Channel<T>>`. And also because we need to have both `channel` (`&Channel`) and `channel_ptr` (`NonNull<Channel>`) in scope simultaneously in a lot of code. So it would be nice to be able to consistently distinguish between the two.

Cassy343 replied:
Resolved!

Cassy343 commented:
Yeah I'll take care of that and implement your latest feedback in the next couple of days - work/life got a little busy in the interim.

faern commented on src/lib.rs:
```diff
@@ -218,44 +261,118 @@ impl<T> Sender<T> {
-                unsafe { channel.take_waker() }.unpark();
+                // ORDERING: Synchronizes with the write of the waker to memory, and prevents the
+                // taking of the waker from being ordered before this operation.
+                fence(Acquire);
```
I have now read and understood the UAF issue you described at the start, as well as your proposed fix to it. It's very clever! Thank you for this. I hope and think we can get this integrated somewhat soon.

Cassy343 replied:
Thanks! It was a pleasure (albeit a head-scratching one) to work on this and get it sorted out :)

faern reviewed (changes requested):
I'd like to merge all bugfix changes that does not change the public API in any way first. And release a patch release of the crate. Then have a separate PR that makes any API changes, for example the fact that async receiving is moved to a separate type.

Cassy343 commented:
So just to clarify, you still want me to squash "Implement latest feedback," "Deduplicate dealloc code," and "Improve unsafe hygiene," and then we'll work on merging the rest, or did you just want to all the changes into the patch then make API/docs changes?

faern commented:
I still want the somewhat unrelated changes to be different PRs. So squashing is not strictly necessary, but they have to come sorted one feature/fix after each other, and preferably different PRs. I think the stuff already done in this PR and discussed in this PR should be at least three PRs:

1. Change `*mut` into `NonNull` and wrapping fields in `UnsafeCell` in order to mitigate one UB. Including all code comment updates, `SAFETY` documentation and everything related to fixing just this UB.
2. Make the changes to the updates of the `Channel::state` field. Including relaxing the memory ordering, changing the values of the states to allow the `fetch_add` trick.
3. Move the async receiving into a separate type (This breaks the API and will be a semver breaking release)

And your unrelated changes to the documentation could also be a separate PR. Such as moving everything out from `lib.rs` into `README.md` and including it in the code with `#include`. In general, keep one PR to one bugfix or feature addition. Or there can be multiple fixes in one PR if they are very related. But here I feel like we are changing a bunch of unrelated things.

Cassy343 commented:
Okay so "Improve unsafe hygiene" through the first "cargo fmt" (https://github.com/faern/oneshot/pull/17/commits/409c49879db717bd9569326bb7d0123c2bf54c1f through https://github.com/faern/oneshot/pull/17/commits/3df9227c6f6a4ce147915f93e3103fbf185f0e53) contain the unsafe hygiene improvements but none of the algorithm/state changes, basically item 1 in your list. All the commits after that are item 2, and I'll make a separate PR for item 3.

Cassy343 commented:
That CI failure is a little strange...I'm not able to replicate that locally

Cassy343 commented:
Looks like that CI failure I mentioned was a fluke. Edited the history to fix a couple of safety comments and CI passed this time. For reference, the `recv_deadline_and_timeout_time_should_elapse` test failed, but I couldn't work out which flags were on for it since the log was kind of messy.

faern commented:
Thanks. Would you like to submit a PR on step 1 (unsafe code hygiene)? Or do you want me to just pull those commits into master? I see that you still have some changes to that code in the later commits. If I pull it in manually I will backport those changes, so they end up where they are supposed to be. For example stuff like this: https://github.com/faern/oneshot/pull/17/commits/1906dda80d16a4a8f35704b0567dc0a6747e63d2#diff-6ff19a1ff611f3835b5a025663cf4dec901ed914626f78847414836439bd1a79R41

faern commented on src/lib.rs:
```diff
@@ -218,44 +259,120 @@ impl<T> Sender<T> {
+        // exclusive access to this memory location to perform this write.
+        unsafe {
+            channel.write_message(message);
+        }
```
If you put the semicolon after the closing bracket, it becomes a single line of code. This is a neat trick that both of us have already used a lot.

Cassy343 commented:
I'll go ahead and put together a separate PR for that. I think there will still be some rebase conflicts either way but I don't really mind dealing with those.

faern commented:
Since #18 is now merged: Feel free to rebase this code to handle the relaxing of atomic constraints + fix the race due to how `state` is accessed here.

Cassy343 commented:
Alright so this is rebased on master now, I'm going to convert this PR into the PR for relaxing the orderings and fixing the parker-related UB.

faern commented on src/lib.rs:
```diff
@@ -294,17 +344,47 @@ impl<T> Sender<T> {
+                // before this.
+                channel.state.store(DISCONNECTED, Release);
+
+                waker.unpark();
```
Can't the unpark be moved up to before the store if the ordering is release? Won't we also need acquire to guarantee we wake the receiver up after we store `DISCONNECTED`? If true, the same issue exists in the `send` method.

Cassy343 replied:
Now I don't have any docs to point to per se - but what I have always been told (and this is also how it's currently implemented) is that `park`/`unpark` come with implied `Acquire`/`Release` semantics, respectively. I wish this was more clearly documented, but the best justification I have for this assertion is that this pattern is duplicated in the [example](https://doc.rust-lang.org/std/thread/fn.park.html#examples) in `std`.

Really the docs for park/unpark should guarantee stronger memory ordering. I'm considering filing an issue for that.

faern replied:
Upon further investigation I come to a different conclusion. Yes, the current implementation of `unpark` has `Release` ordering, making it impossible for the state store to move to after the unpark. However, the documentation for parking states:

> Notice that being unblocked does not imply any synchronization with someone that unparked this thread

The example you point to even rely on the actual fences in the current implementation, and could deadlock if `park`/`unpark` did not actually synchronize. I was going to write an issue about this discrepancy. But then I found the following, someone beat me to it by little under two weeks: https://github.com/rust-lang/rust/pull/99587. So let's follow that PR and see what the libs team say about the issue. Until then we can rely on the current implementation. But if they claim that the docs are correct, we need to insert stronger orderings here in our code to be on the safe side.

Cassy343 replied:
Yeah the docs as they read currently do not make that promise, but the example given in the docs is allowed to deadlock according to the guarantees provided. So that PR is going to need to end in *some* change to the documentation, and as you said I think it's reasonable to assume that it'll guarantee some sort of release ordering.

faern replied:
I realize that we should not be talking about the properties of `std::thread::Thread::unpark()` so much. This method call is on `ReceiverWaker::unpark()`. And that method does **not** have any release ordering guarantees. This unpark will lead to `Waker::wake` if the receiver is receiving anync.

I think we need to add stronger more explicit orderings here no matter how `Thread::unpark` is implemented or what it guarantees.

Cassy343 replied:
That should be fixable by changing that store to a swap with `AcqRel` ordering and simply discarding the return value.

faern commented on src/lib.rs:
```diff
@@ -271,21 +275,67 @@ impl<T> Sender<T> {
             // The receiver is alive and has not started waiting. Send done.
             EMPTY => Ok(()),
             // The receiver is waiting. Wake it up so it can return the message.
+            #[allow(unreachable_code)]
```
I'd prefer another way to work around this issue. My understanding is that this is only a problem when no features are enabled, right? Instead of these blanked-allows that risks hiding future issues I think we should employ someting like this: https://github.com/faern/oneshot/commit/0d3e79175d6b217146040a58ae8c2cdbeb0c12eb

Feel free to cherry-pick this commit into the PR in order to get rid of these `#[allow..]`s

Cassy343 replied:
The problem is that with all features disabled, the warnings became incredibly infectious, bubbling all the way up to where even more imports had to be gated (and that's already messy), and one of the states being unused (`UNPARKING`). I can propagate all those up if you want me too but the end result might not be much better than this already non-ideal solution.

faern replied:
I'm not able to reproduce that. If I cherry-pick my linked commit onto this PR branch and build/test with no features I get no warnings. Neither with nor without loom.

Cassy343 replied:
Oh I glanced at that commit too quickly and didnt see what it was doing. That seems like a reasonable alternative.

faern commented on src/lib.rs:
```diff
@@ -250,7 +255,6 @@ pub struct Receiver<T> {
 
 unsafe impl<T: Send> Send for Sender<T> {}
 unsafe impl<T: Send> Send for Receiver<T> {}
-impl<T> Unpin for Receiver<T> {}
```
This is a breaking change. Before, `Receiver<T>` implemented `Unpin` for all `T`. But now it only does it for `T: Unpin`... Does the message really need to implement `Unpin` for the receiver to do the same? :thinking:

Cassy343 replied:
Good calling that out. I originally removed it since the receiver just contained a nonnull which is always Unpin, but now that the phantom data is there that's no longer the case so the impl needs to be re added

faern commented:
Hi, what's up? I have been away for a while. There are large parts of this PR I have not yet reviewed. But what's the status on fixing some stuff around the things I pointed out ~5 days ago.

Cassy343 commented:
Hi, I'll get to it as soon as I can, I've been really busy these last couple days but hope to have some time this week.

Cassy343 commented on src/lib.rs:
```diff
@@ -846,7 +1072,7 @@ impl<T> Channel<T> {
         match self
             .state
-            .compare_exchange(EMPTY, RECEIVING, SeqCst, SeqCst)
+            .compare_exchange(EMPTY, RECEIVING, AcqRel, Acquire)
```
When you say "motivated," do you mean justifying why it's _not_ `SeqCst`, or justifying why the new ordering is sufficient/correct?

faern replied:
In general I mean adding code documentation justifying the ordering the code uses.

faern commented on src/lib.rs:
```diff
@@ -294,17 +344,48 @@ impl<T> Sender<T> {
+                channel.state.swap(DISCONNECTED, AcqRel);
+
+                // `unpark` has an implicit release ordering, so the store of the DISCONNECTED
+                // state happens-before we unpark the receiver
```
This comment is now outdated. We don't rely on this implicit undocumented release ordering that is supposedly not part of the stable API of `unpark`.

faern commented on src/lib.rs:
```diff
@@ -366,14 +450,25 @@ impl<T> Receiver<T> {
     #[cfg(feature = "std")]
     pub fn recv(self) -> Result<T, RecvError> {
+        // Note that we don't need to worry about changing the state to disconnected or setting the
+        // state to an invalid value at any point in this functon because we take ownership of
```
Nit. s/functon/function/

faern commented on src/lib.rs:
```diff
@@ -383,57 +478,104 @@ impl<T> Receiver<T> {
 
                 // Write our waker instance to the channel.
+                // SAFETY: we are still in the empty state, meaning that the sender will not try
+                // to access the waker until it sees the state set to RECEIVING
```
We can't guarantee that the channel is in the `EMPTY` state when we write the waker here. It could also be `MESSAGE` or `DISCONNECTED` if the `Sender` sends or is dropped in between here. But I don't think you need to explain all of that. It's enough to explain that the `Sender` will not access the waker until it sees the `RECEIVING` state, which we are responsible for setting.

faern commented on src/lib.rs:
```diff
@@ -383,57 +478,104 @@ impl<T> Receiver<T> {
-                    .state
-                    .compare_exchange(EMPTY, RECEIVING, SeqCst, SeqCst)
-                {
+                // Switch the state from EMPTY to RECEIVING. We need to do this in one
```
Same as above. We don't really know that we are in the `EMPTY` state. I think it would be clearer to state that we want to set the state to `RECEIVING` while also fetching the previous state to act on it.

faern commented on src/lib.rs:
```diff
@@ -366,14 +450,25 @@ impl<T> Receiver<T> {
-        match channel.state.load(SeqCst) {
+        // ORDERING: we use acquire ordering to synchronize with the write of the message in the
+        // case that it's available
+        match channel.state.load(Acquire) {
```
What's the motivation for putting the `Acquire` ordering here rather than an explicit `fence(Acquire);` down in the `MESSAGE` branch? Only one branch need this ordering, so why pay for it in all branches?

Cassy343 replied:
There's generally no hard-and-fast rule about when to use a fence vs not. I would classify any potential performance win there as an over-optimization however, because if we end up _not_ needing the acquire barrier, then we will _probably_ end up parking, which will take immensely more wall-clock time. I've also heard that sometimes `load(Acquire)` can be more efficient on some platforms, but I've not been given any specific examples or benchmarks.

Cassy343 replied:
In briefly talking with some other devs, someone claims to have benched that `load(Acquire)` is faster than `load(Relaxed); fence(Acquire);` on aarch64. I think there might be an argument for getting rid of the fences on the sending side given that those branches are likely to be taken. I'd mostly consider using the fence optimization for branches which are unlikely to be taken.

faern replied:
Ok. Yeah, good argument. I was not aware that specifying the ordering in the load could be optimized better. But I guess that makes sense. Would be cool to run those benchmarks or write my own at test it. Do you have any links to anything on this?

faern reviewed (changes requested):
Sorry for the slow reviewing! I want this merged and released as much as you do, I have just been busy with other stuff.

I have now reviewed everything down to (not including) `recv_ref`. From there on and down it seemed to be some extra trickery, so I'll save that for tomorrow.

The feedback I left now is almost only just documentation/comment updates. Everything so far seems sound and really good!

faern commented on src/lib.rs:
```diff
@@ -556,10 +663,117 @@ impl<T> Receiver<T> {
+        &self,
+        disconnected: impl FnOnce() -> E,
+        finish: impl FnOnce(&Channel<T>) -> Result<T, E>,
+    ) -> Result<T, E> {
```
I think some documentation on roughly what `start_recv_ref` is doing would be great for understanding this code.

faern commented on src/lib.rs:
```diff
@@ -556,10 +663,117 @@ impl<T> Receiver<T> {
+    #[inline(always)]
+    fn start_recv_ref<E>(
+        &self,
+        disconnected: impl FnOnce() -> E,
```
In all the places where this is called, the `disconnected` closure does not contain any computation or captured data, it just returns a statically known error value. Can't this be simplified to `disconnected_error: E`?

faern commented on src/lib.rs:
```diff
@@ -455,68 +597,33 @@ impl<T> Receiver<T> {
+                        // We take the message and mark the channel disconnected.
+                        MESSAGE => {
+                            // ORDERING: the sender is inactive at this point so we don't need to make
+                            // and reads or writes visible to the sending thread
```
"to make and reads" should probably be "to make our reads"?

Cassy343 replied:
I think I meant "to make any reads"

faern commented on src/lib.rs:
```diff
@@ -556,10 +663,117 @@ impl<T> Receiver<T> {
+    }
+
+    #[cfg(feature = "std")]
+    #[inline(always)]
```
Are you sure it's beneficial to always inline this? It's a pretty beefy method.

faern commented on src/lib.rs:
```diff
@@ -638,16 +845,19 @@ impl<T> core::future::Future for Receiver<T> {
                 match channel
                     .state
-                    .compare_exchange(RECEIVING, EMPTY, SeqCst, SeqCst)
+                    .compare_exchange(RECEIVING, EMPTY, Relaxed, Relaxed)
```
Can we get a motivation for this ordering please? Or revert it to something stronger. Even if we want to move all the `Future` stuff out to its own type. I need to be able to release a `0.1.4` with all these fixes and the same API first. So we need to make sure the `Future` implementation is sound even before doing the bigger refactor.

faern reviewed (changes requested):
I have now reviewed the rest! I could not stop myself. Too exciting reading material ;)

This is great work! I have some questions and suggestions, but overall I think this is very close to being merged.

Cassy343 commented:
Thanks for putting in so much time to look over all of that! I am also really busy right now so I will get to that over the course of this week.

Cassy343 commented on src/lib.rs:
```diff
@@ -568,63 +782,56 @@ impl<T> Receiver<T> {
+
+                        // ORDERING: the sender has been `mem::forget`-ed so this update only
+                        // needs to be visible to us
+                        channel.state.store(DISCONNECTED, Relaxed);
```
We're allowed to assume that the process of transferring ownership between threads has appropriate acquire/release semantics. So in this particular case that store would be released, and the acquired by the other thread, providing sufficient synchronization.

faern replied:
Ok. I was thinking along those lines as well and assumed that would be the case for most/all common ways of sending the instance over. And I could not come up with any real way of proving otherwise.

Cassy343 commented on src/lib.rs:
```diff
@@ -568,63 +782,56 @@ impl<T> Receiver<T> {
-                    // The sender sent the message while we prepared to park.
+                    // We stored our waker, now we delegate to the callback to finish the receive
+                    // operation
+                    Ok(EMPTY) => finish(channel),
```
I looked at godbolt briefly and it does seem like that slightly helped codegen, I'll change that everywhere that's relevant in the code.

Cassy343 commented:
All feedback should be addressed except that last review there. Once that's finalized I think this is pretty good to go. `IntoFuture` is still on track to land in 1.64 AFAIK, so I think what I might do is just draft up a PR using the beta feature and I don't think it's unreasonable to just wait until September to land that.

faern reviewed (commented):
I aim to review this again/go over the changes during this weekend. Is there anything left you want to do or do you feel it's ready for a (hopefully) final review?

Regarding `IntoFuture`: It does not hurt to have an explicit `fn recv_async(&self) -> AsyncReceiver`. This works on stable today and can be implemented and merged without consideration for MSRV or anything else. Then we can add an `IntoFuture` impl which does the same thing, but implicitly. Depending on when that happens we can decide if it should be behind a feature gate or not.

faern commented on tests/sync.rs:
```diff
@@ -203,6 +203,26 @@ fn try_recv() {
+            mem::drop(receiver);
+        });
+        t1.join().unwrap();
+        t2.join().unwrap();
```
Is this supposed to test the race condition that the library had before this PR? I can't make this test fail even when I backport it to current `master`. Of course a normal run won't trigger it very often at all. But if it really was a thread interleaving issue loom should pick it up, no?

Would be nice to understand the situation and have a test that actually catch the issue. Otherwise it won't really protect against regressions.

Cassy343 replied:
I think this was to test the interleaving of recv fails, send happens, receiver is dropped. I think I added this test while I was working out how memory reclamation worked in this crate since I didn't see a test that tested that ordering.

faern reviewed (approved):
Ok. I'm about to merge this. There is obviously some work left on the `Future` impl before we can publish. But I think we can do that in a separate PR. Mostly because this one is large and can easily have bitrot.

Cassy343 commented:
Thanks! I'm very glad to see this merged. I'll figure out async receiving and draft something up over the next week or two. I'm moving back to college soon so it's a busy time for me

faern commented:
@Cassy343 Thanks for all your hard work and clever solutions! :heart: 

I'm currently looking at the `Future` impl to see if there are any low hanging fruit I can fix right now. Just so we are on the same page. The current impl is unsound, right? It's not only that it has stricter orderings than necessary, right?

Cassy343 commented:
I don't remember off the top of my head if it's unsound or not. If memory serves me correctly then I think there's issues with the writing/unparking of the waker being racy. I think a correct implementation might involve using something like `atomic-waker`, but we might be able to avoid depending on/recreating that since we already keep track of channel state atomically. The tricky part is that we generally need to write the new waker to memory before observing the state, otherwise between the time we observed the state and wrote the waker, the state may have changed causing a race or causing us to write the new waker after the sending party woke the old waker.

faern commented:
I guess upon polling, if we need to store a waker but one is already written, we might need to steal the waker in the same way the `Sender` is doing it: First resetting the state to one where the other entity (in this case the `Sender`) is not allowed to read/drop the waker. If we succeed with that state transition we take out the old waker and drop it properly, then we write the new waker and set the state to `RECEIVING` *again*.

It's going to be quite the dance.

faern commented:
I created a follow up PR that fixes the `Future` implementation as well as some other improvements. See #20

tisonkun commented on src/lib.rs:
```diff
@@ -673,25 +887,31 @@ impl<T> core::future::Future for Receiver<T> {
             // The receiver has been polled.
             #[cfg(feature = "async")]
             RECEIVING => {
+                // TODO: figure this out when async is fixed
```
What's the background of TODO here? I don't find it's still valid now.

Cassy343 replied:
I'm guessing this was a note to think about what would need to happen if the receiver was dropped while in this state. In the sync version this is not possible, since either the receiver is owned or guaranteed to remain alive because of a borrow, but in an async context this is technically possible, although I think the situations in which such a thing would happen and be observable by a program are fairly pathological. It's entirely possible that this is the correct thing to do (and at a glance I'd believe it), but it could use some more careful thought, perhaps.

tisonkun commented on src/lib.rs:
```diff
@@ -656,6 +865,11 @@ impl<T> core::future::Future for Receiver<T> {
                     // The sender has taken the waker already.
                     Err(DISCONNECTED) => Poll::Ready(Err(RecvError)),
+                    Err(UNPARKING) => {
+                        cx.waker().wake_by_ref();
+                        hint::spin_loop();
+                        Poll::Pending
```
If we're in `UNPARKING`, isn't it mean that there is already a message?

We should have directly taken the message here.

cc @Cassy343 @faern WDYT?

tisonkun replied:
Well. This can be from `Drop` Sender, then we may reuse the busy loop to handle UNPARKING else where.

Cassy343 replied:
@tisonkun It's been a while since I've wrapped my head around this, but I believe the rationale is that if the receiver eagerly takes the message, then the receiver cannot safely deallocate the underlying memory since (in principal) the sender could still be accessing the waker. Even if we just eagerly take the message in `poll`, then `Receiver::drop` doesn't know if the message has been taken or not if the `UNPARKING` state is observed. Now, there might be a way to restructure the sending side a bit, and force it to be responsible for deallocation if it observes `DISCONNECTED` [here](https://github.com/faern/oneshot/blob/main/src/lib.rs#L322). This could potentially allow for immediately taking the message, but I could be missing something, I didn't think through it super thoroughly before writing this.
